warehouse/oidc/services.py

```diff
@@ -0,0 +1,181 @@
+        """
+        Returns a reasonable Redis key for storing JWKs for the given provider.
+        """
+        return f"/warehouse/oidc/jwks/{provider}"
+
+    def _provider_timeout_key(self, provider):
+        """
+        Returns a reasonable Redis key for storing the "timeout" flag for the
+        given provider, indicating whether we should attempt to refresh the
+        keyset.
+        """
+        # NOTE: We could use a TTL on the actual Redis key for the keyset,
+        # but only once Warehouse upgrades to Redis >= 6.0.
+        return f"{self._provider_jwk_key(provider)}/timeout"
+
+    def _store_keyset(self, provider, keys):
+        """
+        Store the given keyset for the given provider, setting the timeout key
+        in the process.
+        """
+        with redis.StrictRedis.from_url(self._cache_url) as r:
+            r.set(self._provider_jwk_key(provider), json.dumps(keys))
+            r.setex(self._provider_timeout_key(provider), 60, "placeholder")
+
+    def _get_keyset(self, provider):
+        """
+        Return the cached keyset for the given provider, if it exists,
+        along with the timeout state.
+        """
+        with redis.StrictRedis.from_url(self._cache_url) as r:
+            keys = r.get(self._provider_jwk_key(provider))
+            timeout = bool(r.exists(self._provider_timeout_key(provider)))
+            if keys is not None:
+                return (json.loads(keys), timeout)
+            else:
+                return (None, timeout)
+
+    def _fetch_keyset(self, provider):
+        oidc_url = f"{oidc.OIDC_PROVIDERS[provider]}/{oidc.WELL_KNOWN_OIDC_CONF}"
+
+        resp = requests.get(oidc_url)
+
+        # For whatever reason, an OIDC provider's configuration URL might be
+        # offline. We don't want to completely explode here, since other
+        # providers might still be online (and need updating), so we spit
+        # out an error and return None instead of raising.
+        if not resp.ok:
+            logger.error(
+                f"error querying OIDC configuration for {provider}: {oidc_url}"
+            )
+            return None
+
+        oidc_conf = resp.json()
+        jwks_url = oidc_conf["jwks_uri"]
+
+        resp = requests.get(jwks_url)
+
+        # Same reasoning as above.
+        if not resp.ok:
+            logger.error(f"error querying JWKS JSON for {provider}: {jwks_url}")
+            return None
+
+        jwks_conf = resp.json()
+        keys = jwks_conf["keys"]
+
+        # Another sanity test: an OIDC provider should never return an empty
+        # keyset, but there's nothing stopping them from doing so. We don't
+        # want to cache an empty keyset just in case it's a short-lived error,
+        # so we check here, error, and return None instead.
+        if len(keys) == 0:
+            logger.error(f"{provider} returned JWKS conf but no keys")
+            return None
+
+        return keys
+
+    def get_key(self, provider, key_id):
```

Comment:
I'd like to see this simplified a bit, something like:

```python
    def get_key(self, provider, key_id):
        keyset = self._get_keyset()
        if key_id not in keyset:
            keyset = self._refresh_keyset()
        if key_id not in keyset:
            # emit metrics
            return None
        return PyJWT(keyset[key_id])
```

This function doesn't need to know anything about timeouts or how to store a keyset, which can both be handled by `_refresh_keyset()`. Similarly `_refresh_keyset()` doesn't need to know about what `key_id` we're looking for here, just that the keyset might be stale.

Reply:
Sounds good! That's a much better isolation of concerns; I'll refactor towards this.